W1514 unspecified-encoding false positive when mode is specified using a variable and it can be 'b' (binary mode)

### Bug description

```python
@contextmanager
def open_file_seek_zero(some_str, mode):
    ...
    with open(file_name, mode) as f:        # <===   this line is reported for W1514
        f.seek(0)
        yield f

============ above function usage ============
def some_other_func:
    with open_file_seek_zero(some_str, "r+b") as f:
        ...
==============================================
```

### Configuration

```ini

```

### Command used

```shell
pylint --rc=.pylintrc file_ops.py (made up file name)
```

### Pylint output

```python
************* Module file_ops
file_ops.py:29:9: W1514: Using open without explicitly specifying an encoding (unspecified-encoding)
```

### Expected behavior

When file opening mode is not clearly discernable, unspecified-encoding should not be reported.

If I disable it in rc, I won't catch genuine errors.

### Pylint version

```shell
pylint 3.3.3
astroid 3.3.8
Python 3.10.5
```

### OS / Environment

_No response_

### Additional dependencies

```python

```

--- a/doc/whatsnew/fragments/10201.false_positive
+++ b/doc/whatsnew/fragments/10201.false_positive
@@ -1,0 +1,4 @@
+Fix ``unspecified-encoding`` false positive when an ``open`` call uses a mode
+argument that cannot be inferred.
+
+Closes #10201

--- a/doc/whatsnew/fragments/10201.false_positive
+++ b/doc/whatsnew/fragments/10201.false_positive
@@ -1,4 +1,4 @@
-Fix ``unspecified-encoding`` false positive when an ``open`` call uses a mode
+Fix ``ref:unspecified-encoding`` false positive when an ``open`` call uses a mode
 argument that cannot be inferred.
 
 Closes #10201

--- a/doc/whatsnew/fragments/10201.false_positive
+++ b/doc/whatsnew/fragments/10201.false_positive
@@ -1,4 +1,5 @@
-Fix ``ref:unspecified-encoding`` false positive when an ``open`` call uses a mode
-argument that cannot be inferred.
+Fix a false positive for :ref:`unspecified-encoding` when an ``open`` call uses a mode
+argument that cannot be inferred, including when the mode is a parameter with a default
+value.
 
 Closes #10201

--- a/doc/whatsnew/fragments/10201.false_positive
+++ b/doc/whatsnew/fragments/10201.false_positive
@@ -1,5 +1,4 @@
 Fix a false positive for :ref:`unspecified-encoding` when an ``open`` call uses a mode
-argument that cannot be inferred, including when the mode is a parameter with a default
-value.
+argument that cannot be inferred.
 
 Closes #10201

--- a/pylint/checkers/stdlib.py
+++ b/pylint/checkers/stdlib.py
@@ -887,10 +887,7 @@ class StdlibChecker(DeprecatedMixin, BaseChecker):
 
         if not mode_arg_is_unknown and (
             not mode_arg
-            or (
-                isinstance(mode_arg, nodes.Const)
-                and "b" not in str(mode_arg.value)
-            )
+            or (isinstance(mode_arg, nodes.Const) and "b" not in str(mode_arg.value))
         ):
             confidence = HIGH
             try:

--- a/pylint/checkers/stdlib.py
+++ b/pylint/checkers/stdlib.py
@@ -849,6 +849,7 @@ class StdlibChecker(DeprecatedMixin, BaseChecker):
     ) -> None:
         """Various checks for an open call."""
         mode_arg = None
+        mode_arg_is_unknown = False
         confidence = HIGH
         try:
             if open_module == "_io":
@@ -866,6 +867,10 @@ class StdlibChecker(DeprecatedMixin, BaseChecker):
 
         if mode_arg:
             mode_arg = utils.safe_infer(mode_arg)
+            if func_name in OPEN_FILES_MODE and (
+                mode_arg is None or isinstance(mode_arg, util.UninferableBase)
+            ):
+                mode_arg_is_unknown = True
             if (
                 func_name in OPEN_FILES_MODE
                 and isinstance(mode_arg, nodes.Const)
@@ -880,8 +885,12 @@ class StdlibChecker(DeprecatedMixin, BaseChecker):
                     confidence=confidence,
                 )
 
-        if not mode_arg or (
-            isinstance(mode_arg, nodes.Const) and "b" not in str(mode_arg.value)
+        if not mode_arg_is_unknown and (
+            not mode_arg
+            or (
+                isinstance(mode_arg, nodes.Const)
+                and "b" not in str(mode_arg.value)
+            )
         ):
             confidence = HIGH
             try:

--- a/pylint/checkers/stdlib.py
+++ b/pylint/checkers/stdlib.py
@@ -849,7 +849,6 @@ class StdlibChecker(DeprecatedMixin, BaseChecker):
     ) -> None:
         """Various checks for an open call."""
         mode_arg = None
-        mode_arg_is_unknown = False
         confidence = HIGH
         try:
             if open_module == "_io":
@@ -867,27 +866,21 @@ class StdlibChecker(DeprecatedMixin, BaseChecker):
 
         if mode_arg:
             mode_arg = utils.safe_infer(mode_arg)
-            if func_name in OPEN_FILES_MODE and (
-                mode_arg is None or isinstance(mode_arg, util.UninferableBase)
-            ):
-                mode_arg_is_unknown = True
-            if (
-                func_name in OPEN_FILES_MODE
-                and isinstance(mode_arg, nodes.Const)
-                and not _check_mode_str(mode_arg.value)
-            ):
-                self.add_message(
-                    "bad-open-mode",
-                    node=node,
-                    # avoid a boolean context on the constant ``bool(NotImplemented)``
-                    # raises a TypeError on Python >= 3.14
-                    args=str(mode_arg.value),
-                    confidence=confidence,
-                )
+            if func_name in OPEN_FILES_MODE:
+                if not isinstance(mode_arg, nodes.Const):
+                    return  # mode may be binary - don't guess
+                if not _check_mode_str(mode_arg.value):
+                    self.add_message(
+                        "bad-open-mode",
+                        node=node,
+                        # avoid a boolean context on the constant ``bool(NotImplemented)``
+                        # raises a TypeError on Python >= 3.14
+                        args=str(mode_arg.value),
+                        confidence=confidence,
+                    )
 
-        if not mode_arg_is_unknown and (
-            not mode_arg
-            or (isinstance(mode_arg, nodes.Const) and "b" not in str(mode_arg.value))
+        if not mode_arg or (
+            isinstance(mode_arg, nodes.Const) and "b" not in str(mode_arg.value)
         ):
             confidence = HIGH
             try:

--- a/tests/functional/u/unspecified_encoding_py38.py
+++ b/tests/functional/u/unspecified_encoding_py38.py
@@ -194,3 +194,10 @@ Path(FILENAME).read_text(**KWARGS)  # [unspecified-encoding]
 
 KWARGS = {"encoding": "utf-8"}
 Path(FILENAME).read_text(**KWARGS)
+
+
+def open_with_unknown_mode(mode):
+    """Don't emit unspecified-encoding when the mode might be binary."""
+    open(FILENAME, mode)
+    open(FILENAME, mode=mode)
+    Path(FILENAME).open(mode)

--- a/tests/functional/u/unspecified_encoding_py38.py
+++ b/tests/functional/u/unspecified_encoding_py38.py
@@ -201,4 +201,5 @@ def open_with_unknown_mode(mode):
     open(FILENAME, mode)
     open(FILENAME, mode=mode)
     Path(FILENAME).open(mode)
-    open(f, mode, encoding=None)
+    Path(FILENAME).open(mode=mode)
+    open(FILENAME, mode, encoding=None)

--- a/tests/functional/u/unspecified_encoding_py38.py
+++ b/tests/functional/u/unspecified_encoding_py38.py
@@ -201,3 +201,4 @@ def open_with_unknown_mode(mode):
     open(FILENAME, mode)
     open(FILENAME, mode=mode)
     Path(FILENAME).open(mode)
+    open(f, mode, encoding=None)

